src/fudge/log.clj

```diff
@@ -118,40 +130,62 @@
 
 ;; # Main logging functions
 
-(defn log
-  "Log a record with the config, level, and data provided"
-  [c level data]
-  (let [record (invoke c :prepare-fn data level)]
+(defn log!
+  "Log a record with the config, level, form metadata and data provided"
+  [c calling-meta level data]
+  (let [record (invoke c :prepare-fn data level calling-meta)]
     (when (invoke c :log?-fn record)
       (->> record
            (call c :format-fn)
            (call c :output-fn)))))
 
-(defn spy-with
+(defmacro log
+  "Log a record with the config, level, and data provided."
+  [c level data]
+  `(log! ~c (calling-meta) ~level ~data))
+
+(defmacro spy-with->>
```

Comment:
So it can use the `calling-meta` macro correctly. With two macros like this, the whole lot expands into the caller's namespace, which means `calling-meta` gets the correct file, namespace and line. If this was a function, `calling-meta` would always return the file as `fudge/log.clj`, the namespace as `fudge.log` and the line as whatever line the call to `calling-meta` was on.

It's why I tried to leave all the complicated stuff in the log function, and have each macro do the minimum necessary to pass the function what it needs. Then there's minimal faffing with macro escaping.